### PR TITLE
Add Sphinx documentation (incl. built html files) and deploy workflow

### LINKED ISSUE #10: Update Sphinx Documentation: Docstring Inheritance Fix and Auto-Deployment
Our current Sphinx documentation seems to have incorrect docstring inheritance (not displayed at all), and we should streamline the deployment process by automatically publishing the documentation to e.g. GitHub Pages (for alternatives see [Sphinx guidelines](https://www.sphinx-doc.org/en/master/tutorial/deploying.html)).

Implementation Steps:

Docstring Inheritance:
- Verify that inherited docstrings accurately reflect the functionality of overridden methods and attributes and are compatible with Sphinx.

Auto-deployment to e.g. GitHub Pages:
- Search for accessible hosting platforms and open-source standards.
- Set up an automated process to build and deploy the Sphinx documentation.
- Ensure that the documentation build process is robust and can handle variations in the codebase.
- Consider versioning the documentation to maintain historical records of previous releases.

### PR DESCRIPTION
**Pull request by MarcoHuebner**
Rohversion einer Sphinx Doku, dementsprechend ebenfalls `pyproject.toml` dev-dependencies angepasst

**tbd**:
- (Done) Design (aktuell: Sphinx Standard)
- (Done) Testen des Bauens der Dokumentation in CI/CD (und vllt. lokal)
- "Hosting" z.B. auf [readthedocs](https://docs.readthedocs.io/en/stable/tutorial/index.html) -> CorrelAid Account?
  - Automatisches Bauen der Dokumentation bei Code Updates/ [PRs](https://docs.readthedocs.io/en/stable/tutorial/index.html#trigger-a-build-from-a-pull-request)?
- Copyright (CorrelAid...?), Autoren und weitere Metadaten?

resolves #10

### CODE REVIEW

**Comment by pmayd** [comment]
Das sind ne Menge roter Tests XD Schön, dass hier noch weiter gearbeitet wird, mir war nicht bewusst, das jemand daran arbeitet aber ich schaue es mir gerne an!

**Comment by MarcoHuebner** [comment]
Oje - werde mir in den nächsten zwei Wochen mal anschauen, woher die grundlegenden Fehler kommen. Die Rohversion sollte eigentlich keine Tests - außer vllt. code quality - beeinflussen, ich wollte mit dem PR erstmal mögliche feature-Wünsche abklappern :)

**Comment by pmayd** [comment]
Relativ einfach die angegebenen Versionen von Python existieren wohl nicht mehr, warum auch immer, sollte also schnell behoben sein

**Comment by MarcoHuebner** [comment]
Current issues seem to be related to changes/ issues in [urllib3](https://github.com/espressif/idf-component-manager/issues/31#issuecomment-1535984388). Working on a fix.

File: README.md [hunk 1]
```diff
@@ -138,7 +138,6 @@ Distributed under the MIT License. See `LICENSE.txt` for more information.
 A few ideas we should implement in the maybe-near future:
 
 - Improve Table parsing. Right now, the parsing is really simple and we should align the cube and table format so that the data frame for tables is more convenient to use.
-- Create a source code documentation with Sphinx or similar tools.
```
Comment (pmayd) [inline README.md:141]: nice!

File: src/pystatis/cube.py [hunk 2]
```diff
@@ -105,9 +105,8 @@ def rename_axes(
     Args:
         cube (dict): A dictionary holding the cube data as returned by `parse_cube()`.
         rename_classifying_variables (bool, optional): If True, rename classifying variables.
-            Defaults to True.
-        rename_time_variable (bool, optional): If True, rename the time variable.
-            Defaults to True.
+        Defaults to True.
```
Comment (pmayd) [inline src/pystatis/cube.py:108]: not sure about this one but I think Google stylesheet says to intend the line if the string belongs to the previous line so Defaults to True is part of the parameter `rename_classifying_variables`. But not important and if it works with Sphinx fine

**Comment by pmayd** [comment]
> Rohversion einer Sphinx Doku, dementsprechend ebenfalls `pyproject.toml` dev-dependencies angepasst
> 
> **tbd**:
> 
>     * (Done) Design (aktuell: Sphinx Standard)
> 
>     * (Done) Testen des Bauens der Dokumentation in CI/CD (und vllt. lokal)
> 
>     * "Hosting" z.B. auf [readthedocs](https://docs.readthedocs.io/en/stable/tutorial/index.html) -> CorrelAid Account?
>       
>       * Automatisches Bauen der Dokumentation bei Code Updates/ [PRs](https://docs.readthedocs.io/en/stable/tutorial/index.html#trigger-a-build-from-a-pull-request)?
> 
>     * Copyright (CorrelAid...?), Autoren und weitere Metadaten?

yea we could definitely publish to readthedocs, but let's first have the local version working and deployed

File: src/pystatis/cube.py (hunk 2 above)
Comment (MarcoHuebner) [reply src/pystatis/cube.py:108]: Another alternative would be to have it (`Defaults...`) all in one line. At least locally this solution complies with black & pylint. The problem with sphinx is, that the indentations (without clear, return-separated blocks) are "unexpected" (warning, no error).

**Comment by MarcoHuebner** [comment]
Should I add more info on sphinx in the ReadMe? Testing whether the build works is already integrated into the CI/CD (afaik/ afai can test), and local deployment - as far as I understand - would be `cd docs && make html`. That could be clarified/ stated in the ReadMe since I did not do so already.

Opinion/ vision for aforementioned local deployment?

**Comment by MarcoHuebner** [comment]
- Updated ReadMe with a "full documentation" subsection
- Added custom pre-commit hook (but thus had to change the docs package version to a hard-coded one to avoid importing the `pystatis` package. [ReadTheDocs](https://github.com/readthedocs/readthedocs.org/blob/37f9806a0f96f281a2dcd0ed73b930781fa7938a/docs/conf.py#L71) also seems to do it that way... therefore added another version test...
- Uploaded docs as artifact - are the standard 90 days [storage time](https://docs.github.com/de/actions/using-workflows/storing-workflow-data-as-artifacts#configuring-a-custom-artifact-retention-period) ok?

**Comment by MarcoHuebner** [comment]
Detailed walkthrough: https://pages.github.com/

TODOs:
- Create `gh-pages` branch
- Add `index.html` and maybe `.nojekyll` to publish branch
- Add [deploy SSH Key](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-create-ssh-deploy-key) -> requires owner permissions
- set up [GitHub Pages in the settings](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-first-deployment-with-github_token)
- After that, the documentation should automatically be hosted at http://correlaid.github.io/pystatis

**Comment by MarcoHuebner** [comment]
> How is this PR related to #4 ?

Did I write that in a commit message here? If so sorry, the T9-branch and #9 have addressed #4 IMO

**Comment by pmayd** [comment]
I updated the dev branch so you can rebase/merge again.

Is this PR ready for merge?

**Comment by MarcoHuebner** [comment]
> Is this PR ready for merge?

No, since it already has a deploy workflow - which will likely fail without the existence of a `gh-pages` branch to push the built documentation to. If we want to merge it now and reopen another branch later, we would have to remove at least the `deploy-docs.yaml`

**Comment by MarcoHuebner** [comment]
Managed to remove manual updating of the `.rst` files to only having to update the headers and references to the rendered section of the `README.md` (and package metadata).

One minor flaw remains: Images are not rendered correctly (yet).

**Comment by MarcoHuebner** [comment]
The documentation can already be found under Actions / Workflow Runs / <Newest Run> / Artifacts -> docs. This makes the documentation directory downloadable as a `.zip` which can be viewed locally after unpacking.

Next steps:
1. Adjust the artifact storage time if necessary
2. Set up GitHub pages. Steps
- Create a `gh-pages` branch (or any other name)
- Add an `index.html` (copy from build/html) and maybe `.nojekyll` or `_config.yml` (see [here](https://docs.github.com/en/pages/quickstart)) to the documentation publish (e.g. `gh-pages`) branch
- Decide whether to use [GitHub Token](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-first-deployment-with-github_token) or [deploy SSH Key](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-create-ssh-deploy-key) -> 
- After that, the documentation should automatically be hosted at http://correlaid.github.io/pystatis

**Comment by pmayd** [comment]
I have done the necessary setup steps for the deploy SSH key, you can use it now as it is a repository secret:
```yaml

- name: Deploy
  uses: peaceiris/actions-gh-pages@v3
  with:
    deploy_key: ${{ secrets.ACTIONS_DEPLOY_KEY }}
    publish_dir: ./public
```

**Comment by MarcoHuebner** [comment]
Alright, I updated the deploy key now. Things to consider - maybe in the next meeting on Tuesday:
- When should this action be triggered? Only on push/ merge to main? Already on the pull request? What about dev?
- Should we use `force_orphan: true` to only keep the newest documents
- Maybe we'll need `enable_jekyll: false` if we experience problems with file/ folder names starting with a underscore `_`

**Comment by pmayd** [comment]
> Alright, I updated the deploy key now. Things to consider - maybe in the next meeting on Tuesday:
> - When should this action be triggered? Only on push/ merge to main? Already on the pull request? What about dev?
> - Should we use `force_orphan: true` to only keep the newest documents
> - Maybe we'll need `enable_jekyll: false` if we experience problems with file/ folder names starting with a underscore `_`

Thanks! Have to test it out but it sounds perfect.

 No, main is to limited, at least for dev and main you have to check the docu before deploying it to main. However, dev should be enough as a PR against dev already triggers the workflow so I always use PR against dev and main as trigger and I think that is perfect for this, too.

**Comment by MarcoHuebner** [comment]
The first deployment attempt failed with `Error: Action failed with "not found deploy key or tokens"`. If I understand correctly, the SSH Key is personal, so I (and everyone else) would have to set up their own key.
A `GITHUB_TOKEN` however, should allow to automatically authenticate the runner/ workflow run: _["A GitHub Actions runner automatically creates a GITHUB_TOKEN secret to use in your workflow. You can use the GITHUB_TOKEN to authenticate in a workflow run."](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-create-ssh-deploy-key)_

**Comment by pmayd** [comment]
That should not be the case the deploy key is not created under user settings but repo settings and it is a repository secret so no

**Comment by MarcoHuebner** [comment]
True... any obvious errors I'm making in [calling the key](https://github.com/CorrelAid/pystatis/blob/203a56a8e3e2bc7cfe90d565f21d0784f608dc8c/.github/workflows/deploy-docs.yaml#L60)? Typos in `with: deploy_key: ${{ secrets.ACTIONS_DEPLOY_KEY }}`? 🤔

**Comment by pmayd** [comment]
That should not be the case the deploy key is not created under user settings but repo settings and it is a repository secret so no .

From the docu: If the user who created the deploy key is removed from the repository, the deploy key will still be active as it isn't tied to the specific user, but rather to the repository.

**Comment by pmayd** [comment]
I don't think so, according to the doc it should be fairly simple: create the ssh key as public and the secret as private key and that's it...I will have a look later

**Comment by pmayd** [comment]
I found the problem which is logical if you think about it: you are on your own fork and that does not trigger the REPO secrets, you have to call from a branch in this repo.

Here the docu:
"Anyone with collaborator access to this repository can use these secrets and variables for actions. They are not passed to workflows that are triggered by a pull request from a fork."

**Comment by MarcoHuebner** [comment]
Moved this to #27

**Pull request closed without merging**